### PR TITLE
Updated all generate.cmd scripts (WIP)

### PR DESCRIPTION
**Pull request by olydis**
I've used Sergey's information and that of former `generate.cmd`'s to create updated versions that use recent AutoRest and the reorganized repository. As far as I can tell, regeneration successfully recreates the RPs modulo: 1) Changes done in the Swagger since last regeneration; 2) changes in AutoRest since the version that was used last time

Maybe we should rename (instead of overwrite) the existing `generate.cmd`'s so people can fall back to it in case they really rely on an older (Nuget) version of AutoRest.
@shahabhijeet what do you think?

I've made the source repo and branch/commit customizable using CLI arguments: `generate.cmd olydis my-fix` would generate from `olydis/my-fix`, commit IDs naturally work as well. Default is `Azure/current`.
So I guess the guidance to partners is: Use the new script, it should be fine for 90% of them (delete the legacy script in that case). Otherwise we can try adjusting the new script to fit their needs and only fall back to the old script if it must really be an ancient version of AutoRest (although I'd be very interested to hear *why* that is, it would most likely indicate that something is missing in current AutoRest).

Code generation settings are now persisted right in the specs repo, e.g. [here](https://github.com/Azure/azure-rest-api-specs/blob/current/specification/network/resource-manager/readme.md#c), note for example
``` yaml
output-folder: $(csharp-sdks-folder)/Network/Management.Network/Generated
```
which knows where in the SDK repo generated files should go.
This way, generation is very consistent and can be automated more and more (in the future, for example as part of a PR to the specs repo).

### CODE REVIEW

**Comment by shahabhijeet** [comment]
@olydis I have added few folks from various RPs.

**Comment by shahabhijeet** [comment]
@olydis regarding falling back to older nuget for AutoRest. Lets see if there is a push back, if any RP still needs older AutoRest version, I would like to know it, let's find out and we can do whatever is need to unblock.

File: tools/generate.cmd
Comment (jaredmoo) [inline tools/generate.cmd]: please consider adding some kind of `--help` argument so I can remember what param order to use

File: tools/regenerateAll.bat
Comment (jaredmoo) [inline tools/regenerateAll.bat]: Azure/current?

File: tools/regenerateAll.bat
Comment (jaredmoo) [inline tools/regenerateAll.bat]: what do these comments all mean?

**Comment by begoldsm** [comment]
@olydis as a general question for process, do we want to move away from the pattern of forcing the generate.cmd to reference a specific swagger commit, to ensure that it is easy to see the referenced swagger PR that resulted in that commit? I am fine with allowing generation from Azure/Current, just wanted to make sure that, from the Azure SDK perspective, this is intended.

File: src/SDKs/DataLake.Store/Management.DataLake.Store/generate.cmd
Comment (begoldsm) [inline src/SDKs/DataLake.Store/Management.DataLake.Store/generate.cmd]: minor: put in a new line at the end of the file (it was missing previously but we should standardize the empty new line since you are touching the file)

**Review by begoldsm — APPROVED** [review]
Just one comment and the general question, otherwise this looks fine to me.

File: tools/generate.cmd
Comment (olydis) [inline tools/generate.cmd]: will do!

File: tools/regenerateAll.bat
Comment (olydis) [inline tools/regenerateAll.bat]: this file is rather temporary and for my own testing - @shahabhijeet not sure whether it also makes sense otherwise. If so, I'll if course adjust this!

File: tools/regenerateAll.bat
Comment (olydis) [inline tools/regenerateAll.bat]: ...my testing... I manually went through *all RPs* and did a rudimentary diff to see if I broke stuff. The thing is: For some RPs, both the Swagger and the AutoRest version have changed quite a bit, so I'm not always 100% sure the regeneration script is correct.

**Comment by olydis** [comment]
> do we want to move away from the pattern of forcing the generate.cmd to reference a specific swagger commit

@begoldsm yes. Some RP haven't followed the pattern rigorously, others forgot to update the scripts, etc. so instead, we're planning to emit metadata like the commit ID used for generation *as part of the generated files*. So the script can be left untouched, the information is recorded and unless you do extra work, this information also makes it into your PR automatically.

Note however that this requires AutoRest support, so I'll probably squeeze something like that into the `tools/generate.cmd` until then.

File: tools/generate.cmd
Comment (olydis) [inline tools/generate.cmd]: done

File: src/SDKs/DataLake.Store/Management.DataLake.Store/generate.cmd
Comment (olydis) [inline src/SDKs/DataLake.Store/Management.DataLake.Store/generate.cmd]: done

**Comment by olydis** [comment]
@shahabhijeet For generated metadata, what about:
``` Haskell
1) azure-rest-api-specs repository information
GitHub user: olydis
Branch:      adl_object_refactor_rebased
Commit:      b3a8dba9595ecc4175117da3e890ff10ecc6e1a5

2) AutoRest information
Requested version:  latest
Installation:
---------------------------------------------------------------------------
AutoRest Code Generator
(c) 2017 Microsoft Corporation. https://aka.ms/autorest
Build Information
  Autorest Bootstrapper :  1.1.0
  NetCore framework :      1.0.5
  Latest Core Installed :  1.2.2
  Requested Core Version : <none>
Installed versions of AutoRest :

  1.2.2
  1.2.1
  1.1.0
---------------------------------------------------------------------------
```

The first part uniquely identifies the Swagger(s) used for code generation, the second part uniquely identifies the AutoRest version requested and used. (which currently also ends up in all generated files, which we can then stop doing as well)

File: tools/generate.cmd
Comment (jaredmoo) [inline tools/generate.cmd]: You are using some awesome cmd logic that I don't understand. What is the syntax for help here? Seems like `generate help` || `generate ?`, is that right?

**Review by jaredmoo — APPROVED** [review]

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search/Fix-GeneratedCode.ps1
Comment (brjohnstmsft) [inline src/SDKs/Search/DataPlane/Microsoft.Azure.Search/Fix-GeneratedCode.ps1]: Were these somehow accidentally deleted before?

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search/generate.cmd
Comment (brjohnstmsft) [inline src/SDKs/Search/DataPlane/Microsoft.Azure.Search/generate.cmd]: We are using a specific version of AutoRest for the DataPlane because the first non-NuGet versions had some issues that we couldn't work around. We haven't had the time to kick the tires on newer versions since then. It's important that we control the pace of these updates, because we have a lot of hand-written code right now that has to track changes in the AutoRest-generated code.

Please revert this until my team has the bandwidth to take these changes.

**Review by bitsforthought — CHANGES_REQUESTED** [review]
If it's possible to omit the Search DataPlane SDK from these changes, we ask that you do so for now. See my other comments for details.

File: tools/generate.cmd
Comment (shahabhijeet) [inline tools/generate.cmd]: @olydis Usage: needs to also print what argument passed will print Usage?

File: tools/generate.cmd
Comment (shahabhijeet) [inline tools/generate.cmd]: @olydis what is the purpose of providing user info?

File: tools/generate.cmd
Comment (shahabhijeet) [inline tools/generate.cmd]: @olydis currently *.txt that get's generated in addition to the generated code, why do we need this file?
Is this temporary workaround until we have this https://github.com/Azure/autorest/issues/2434?

File: tools/regenerateAll.bat
Comment (shahabhijeet) [inline tools/regenerateAll.bat]: @olydis this seems like a test for making sure generation works?
If that is true? at what point SDK branch should run this test?

**Review by shahabhijeet — CHANGES_REQUESTED** [review]

File: tools/regenerateAll.bat
Comment (olydis) [inline tools/regenerateAll.bat]: it somewhat is a test, but it requires manual inspection of the results (`git diff`)

File: tools/generate.cmd
Comment (olydis) [inline tools/generate.cmd]: @jaredmoo you assume correctly, it's actually anything containing `?` or `help` (I'll slightly adjust it, `%1` is the RP name which could at some point contain `help`, who knows... on the other hand, I wanted the help to be shown if you go `/help` or `-help` or `--help`, so I gotta find some middle ground)
@shahabhijeet will do

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search/Fix-GeneratedCode.ps1
Comment (olydis) [inline src/SDKs/Search/DataPlane/Microsoft.Azure.Search/Fix-GeneratedCode.ps1]: not sure if it was accidentally, just moved the commands from the old `generate.cmd` into this fixup script. Could well be that this is necessary if both Swaggers reference the same types...

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search/generate.cmd
Comment (olydis) [inline src/SDKs/Search/DataPlane/Microsoft.Azure.Search/generate.cmd]: sure, we won't force this update on teams, just offer it - for some teams, the specs-repo reorg makes the old scripts not work anymore (no more composite files) so I wanted to offer a quick solution by migrating everything over. Search data-plane isn't affected by this as far as I can tell, so the old script (with the URIs adjusted to point to the reorged Swaggers) thould totally keep working.

File: tools/generate.cmd
Comment (olydis) [inline tools/generate.cmd]: you mean the GitHub user? So people can use the script to generate code from their own spec reop branches/PRs before they are checked into `Azure/azure-rest-spec-apis`. So essentially this enables testing Swagger changes.

File: tools/generate.cmd
Comment (olydis) [inline tools/generate.cmd]: > Is this temporary workaround until we have this Azure/autorest#2434?

Yes - although, as mentioned in standup, I'm not entirely convinced anymore that AutoRest should even do that. Generating this metadata file is super specific to this use of AutoRest. So why not just implement the metadata generation right here? I'm open to discussions ;)

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search/Fix-GeneratedCode.ps1
Comment (brjohnstmsft) [inline src/SDKs/Search/DataPlane/Microsoft.Azure.Search/Fix-GeneratedCode.ps1]: These fixups are only needed when generating SearchIndexClient.

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search/generate.cmd
Comment (brjohnstmsft) [inline src/SDKs/Search/DataPlane/Microsoft.Azure.Search/generate.cmd]: Sounds good

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search/Fix-GeneratedCode.ps1
Comment (olydis) [inline src/SDKs/Search/DataPlane/Microsoft.Azure.Search/Fix-GeneratedCode.ps1]: oooohhhh you literally generate these separately? I somehow assumed this was a workaround for composite Swagger. But yeah, the fact that you had two instead of one combined generate script should've made me wonder 😄

Comment (brjohnstmsft) [inline src/SDKs/Search/DataPlane/Microsoft.Azure.Search/Fix-GeneratedCode.ps1]: They are separate because we actually want two different clients. Our programming model works like this:

    var index = new Index() { Name = "test", ... };
    var serviceClient = new SearchServiceClient(...);
    serviceClient.Indexes.Create(index);
    ISearchIndexClient indexClient = serviceClient.Indexes.GetClient("test");
    var results = indexClient.Documents.Search(...);

Every index has its own schema, so we have overloads of `Search` and friends that take a generic type parameter for a customer-supplied model class, and other overloads that just return property bags (`Dictionary<string, object>` more or less).

Eventually we hope that `SearchServiceClient` can be 100% generated code, but `SearchIndexClient` will never be.

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search/generate.cmd
Comment (olydis) [inline src/SDKs/Search/DataPlane/Microsoft.Azure.Search/generate.cmd]: @brjohnstmsft BTW, would be good to know which issues you found, we should resolve those. We've moved a long way since the NuGet versions regarding Swagger features but also *flexibility* of code generation. Would be cool if we had moved everybody over, I'd be glad to help.

**Review by bitsforthought — APPROVED** [review]

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search/generate.cmd
Comment (brjohnstmsft) [inline src/SDKs/Search/DataPlane/Microsoft.Azure.Search/generate.cmd]: @olydis Thanks! I'll make sure to sync up with you the next time we make a push on this, but it's probably not going to be until at least October.

**Comment by olydis** [comment]
@shahabhijeet Any update on this? What's your migration plan?

File: tools/regenerateAll.bat
Comment (shahabhijeet) [inline tools/regenerateAll.bat]: @olydis what I am still not clear is, what is the purpose of this script if I am not going to use it?
Is this more appropriate in the spec repo?

**Review by shahabhijeet — CHANGES_REQUESTED** [review]
Let's get this checked in, the more time we wait the more conflicts this PR would have.
I was waiting for the CI to pass and hence the PR was ignored.
Also I think the regenerateAll.bat is more appropriate in spec repo.

File: tools/regenerateAll.bat
Comment (olydis) [inline tools/regenerateAll.bat]: well, *I* was actually wondering whether you may have use for the script. It makes no sense in the specs repo, I thought maybe you have value for such a script (automated tests or similar) - will remove!

---
In reply to: [129956036](https://github.com/Azure/azure-sdk-for-net/pull/3489#discussion_r129956036) [](ancestors = 129956036)

**Pull request merged**
